### PR TITLE
Include NFSTurbo client along with mount helper .deb package

### PR DESCRIPTION
**Pull request by shubham808**

### CODE REVIEW

File: turbonfs/CMakeLists.txt [hunk 1]
```diff
@@ -56,14 +56,21 @@ if(ENABLE_CHATTY)
   add_definitions(-DENABLE_CHATTY)
 endif()
 
+find_program(MESON_EXECUTABLE meson REQUIRED)
+find_program(NINJA_EXECUTABLE ninja REQUIRED)
+
```
Comment (linuxsmiths) [inline turbonfs/CMakeLists.txt:61]: This assumes that meson and ninja are already installed.
Can we instruct it to install the programs if not found. This makes cmake self sufficient which is very important. AT any time I want cmake to be able to build aznfsclient after doing "git clone". No other step should be required.
See how we install tcmalloc if not found.

File: .github/workflows/prerelease.yml [hunk 2]
```diff
@@ -32,11 +32,12 @@ jobs:
             echo "Test Release version must be of the form 0.y.z" 
             false
         fi
-    - name: Install arm64 Cross-Compile Toolchain
+    - name: Install arm64 Cross-Compile Toolchain, meson, ninja-build
```
Comment (linuxsmiths) [inline .github/workflows/prerelease.yml:35]: See my comment in the CMakeLists.txt file. Better to not have to install additional programs, cmake should install if not already installed. That way anyone building from source also can build.

File: package.sh [hunk 3]
```diff
@@ -176,6 +176,13 @@ cp -avf ${SOURCE_DIR}/src/aznfswatchdogv4 ${STG_DIR}/deb/${pkg_dir}/usr/sbin/
 mkdir -p ${STG_DIR}/deb/${pkg_dir}/sbin
 gcc -static ${SOURCE_DIR}/src/mount.aznfs.c -o ${STG_DIR}/deb/${pkg_dir}/sbin/mount.aznfs
 
+pushd ${SOURCE_DIR}/turbonfs
+./build.sh
```
Comment (linuxsmiths) [inline package.sh:180]: By default we should generate release build, but it'll be good to take build_type option from the user like we take for release number so that user can choose release or debug build.

File: package.sh [hunk 4]
```diff
@@ -176,6 +176,13 @@ cp -avf ${SOURCE_DIR}/src/aznfswatchdogv4 ${STG_DIR}/deb/${pkg_dir}/usr/sbin/
...
+popd
+
+mkdir -p ${STG_DIR}/deb/${pkg_dir}/lib
```
Comment (linuxsmiths) [inline package.sh:183]: are you installing anything in lib currently?

File: packaging/aznfs/RPM/aznfs.spec [hunk 5]
```diff
@@ -15,7 +15,7 @@ Mount helper program for correctly handling endpoint IP address changes for Azur
 
 %prep
 mkdir -p ${STG_DIR}/RPM_DIR/root/rpmbuild/SOURCES/
-tar -xzvf ${STG_DIR}/AZNFS_PACKAGE_NAME-${RELEASE_NUMBER}-1.x86_64.tar.gz -C ${STG_DIR}/RPM_DIR/
+tar -xzvf ${SOURCE_DIR}/AZNFS_PACKAGE_NAME-${RELEASE_NUMBER}-1.x86_64.tar.gz -C ${STG_DIR}/RPM_DIR/
```
Comment (linuxsmiths) [inline packaging/aznfs/RPM/aznfs.spec:18]: why is this change needed?

File: turbonfs/CMakeLists.txt [hunk 6]
```diff
@@ -91,28 +98,19 @@ if(NOT EXISTS "${PROJECT_SOURCE_DIR}/extern/spdlog/CMakeLists.txt")
...
-    message(STATUS "Using fuse3 include dir ${fuse3_INCLUDE_DIR}")
-endif()
+if(NOT EXISTS "${PROJECT_SOURCE_DIR}/extern/libfuse/meson.build")
+    message(FATAL_ERROR "The libfuse submodule was not downloaded! GIT_SUBMODULE was turned off or failed. Please update submodules and try again.")
 endif()
 
+add_custom_command(
+    OUTPUT ${LIBFUSE_BUILD_DIR}/lib/libfuse3.a
+    COMMAND ${MESON_EXECUTABLE} setup ${LIBFUSE_BUILD_DIR} ${LIBFUSE_SOURCE_DIR} --default-library=static
+    COMMAND ${MESON_EXECUTABLE} compile -C ${LIBFUSE_BUILD_DIR}
```
Comment (linuxsmiths) [inline turbonfs/CMakeLists.txt:108]: how do we tell it to make release build i.e. with highest compiler optimization. 
Similarly for build_type=debug we should build this with debug options so we can debug in gdb

File: packaging/aznfs/RPM/aznfs.spec (hunk 5 above)
Comment (shubham808) [reply packaging/aznfs/RPM/aznfs.spec:18]: i used different source_dir and stg_dir and this was breaking, i thought this is missed and i fixed it

File: turbonfs/CMakeLists.txt (hunk 6 above)
Comment (shubham808) [reply turbonfs/CMakeLists.txt:108]: Available build types:
debug: Generates debug information but doesn't optimize the code. This is the default.
debugoptimized: Generates debug information and optimizes the code (usually with -g -O2 compiler flags).
release: Optimizes the code without generating debug information.

File: package.sh [hunk 7]
```diff
@@ -176,6 +176,16 @@ cp -avf ${SOURCE_DIR}/src/aznfswatchdogv4 ${STG_DIR}/deb/${pkg_dir}/usr/sbin/
 mkdir -p ${STG_DIR}/deb/${pkg_dir}/sbin
 gcc -static ${SOURCE_DIR}/src/mount.aznfs.c -o ${STG_DIR}/deb/${pkg_dir}/sbin/mount.aznfs
 
+# We build the turbonfs project here, note that we can set all cmake options in the 
+# future using env variables.
+pushd ${SOURCE_DIR}/turbonfs
+mkdir -p build && cd build
+cmake -DCMAKE_BUILD_TYPE=${BUILD_TYPE} -DENABLE_TCMALLOC=OFF ..
```
Comment (linuxsmiths) [inline package.sh:183]: we don't want to turnoff tcmalloc for release builds.
For debug build we do it because debug build also has ASAN enabled and tcmalloc has some know issues working with ASAN. Note that tcmalloc provides better performance and we need it for release builds.

File: packaging/aznfs/RPM/aznfs.spec (hunk 5 above)
Comment (linuxsmiths) [reply packaging/aznfs/RPM/aznfs.spec:18]: You mean you used in your testing? because the PR doesn't have that.
And why was it breaking? STG_DIR is correct here, as package,sh creates the tarball in STG_DIR, as that's a temporary artifact and not the original source code.
We should revert this and if something is breaking, let Palash know so that he can fix it in the correct repo/branches

File: turbonfs/CMakeLists.txt [hunk 8]
```diff
@@ -219,10 +248,8 @@ target_include_directories(${CMAKE_PROJECT_NAME}
                            PRIVATE "${PROJECT_SOURCE_DIR}/extern/spdlog/include")
 
 if(NOT ENABLE_NO_FUSE)
-target_include_directories(${CMAKE_PROJECT_NAME}
-                           PRIVATE "${fuse3_INCLUDE_DIR}")
 target_link_libraries(${CMAKE_PROJECT_NAME}
-                      ${fuse3_LIBRARY})
+                      ${LIBFUSE_BUILD_DIR}/lib/libfuse3.a)
 endif()
```
Comment (linuxsmiths) [inline turbonfs/CMakeLists.txt:253]: instead of removing/changing these I'd suggest we set these variable (fuse_*) correctly after the fuse build+install directions above.
Note that as per the default install dir,
fuse3_LIBRARY will be set to /usr/local/lib/x86_64-linux-gnu, and
fuse3_INCLUDE_DIR will be set to /usr/local/include

fuse3_LIBRARY  is optional as we can directly link from ${LIBFUSE_BUILD_DIR}/lib/libfuse3.a, but for consistency I'd suggest we use that as well.
Then wherever we use ${LIBFUSE_BUILD_DIR}/lib/libfuse3.a we use ${fuse3_LIBRARY}

File: .gitmodules [hunk 9]
```diff
@@ -7,3 +7,6 @@
 [submodule "turbonfs/extern/spdlog"]
 	path = turbonfs/extern/spdlog
 	url = https://github.com/gabime/spdlog.git
+[submodule "turbonfs/extern/libfuse"]
+	path = turbonfs/extern/libfuse
+	url = https://github.com/libfuse/libfuse.git
```
Comment (linuxsmiths) [inline .gitmodules:12]: this comment doesn't apply here but just adding as a placeholder.
Once this change is submitted we should force libfuse to a stable commit not tot.
The last one I see is fuse-3.16.2

**Review by linuxsmiths — COMMENTED** [review]
LG. Let's meet once and close these.

**Review by linuxsmiths — APPROVED** [review]
